src/Http/Http.Abstractions/src/Extensions/UseMiddlewareExtensions.cs

```diff
@@ -215,13 +216,55 @@ public RequestDelegate CreateMiddleware(RequestDelegate next)
             methodArguments[0] = context;
             for (var i = 1; i < parameters.Length; i++)
             {
-                methodArguments[i] = GetService(serviceProvider, parameters[i].ParameterType, methodInfo.DeclaringType!);
+                var parameter = parameters[i];
+
+                var hasServiceKey = TryGetServiceKey(parameter, out object? key);
+                var parameterType = parameter.ParameterType;
+                var declaringType = methodInfo.DeclaringType;
+
+                methodArguments[i] = hasServiceKey ? GetKeyedService(serviceProvider, key!, parameterType, declaringType!) : GetService(serviceProvider, parameterType, declaringType!);
             }
 
             return (Task)methodInfo.Invoke(middleware, BindingFlags.DoNotWrapExceptions, binder: null, methodArguments, culture: null)!;
         };
     }
 
+    private static bool TryGetServiceKey(ParameterInfo parameterInfo, out object? key)
+    {
+        if (parameterInfo.CustomAttributes != null)
+        {
+            foreach (var attribute in parameterInfo.GetCustomAttributes(true))
+            {
+                if (attribute is FromKeyedServicesAttribute keyed)
+                {
+                    key = keyed.Key;
+                    return true;
+                }
+            }
+        }
+        key = null;
```

Comment:
Nit: insert blank line
```suggestion
        }
        
        key = null;
```